udpa/core/v1/resource_locator.proto

```diff
@@ -0,0 +1,97 @@
+//
+// Resource locators also have a simplified file:// URI representation:
+//
+//   file:///{id/*}{#directive*}
+//
+message ResourceLocator {
+  enum Schema {
+    UDPA = 0;
+    HTTP = 1;
+    FILE = 2;
+  }
+
+  // URI schema.
+  Schema schema = 1 [(validate.rules).enum = {defined_only: true}];
+
+  // Opaque identifiers for the resource. These are effectively concatenated
+  // with ‘/’ to form the non-query param path as resource ID. This may end
+  // with ‘*’ for glob collection references.
+  repeated string id = 2 [(validate.rules).repeated = {min_items: 1}];
+
+  // Logical authority for resource (not necessarily transport network address).
+  // Authorities are opaque in the UDPA API, data-plane load balancers will map
+  // them to concrete network transports such as an xDS management server, e.g.
+  // via envoy.config.core.v3.ConfigSource.
+  string authority = 3 [(validate.rules).string = {min_len: 1}];
+
+  // Qualified resource type (as in type URL without types.googleapis.com/
+  // prefix).
+  string qualified_type = 4 [(validate.rules).string = {min_len: 1}];
+
+  oneof context_param_specifier {
+    // Additional parameters that can be used to select resource variants.
+    // Matches must be exact, i.e. all context parameters must match exactly and
+    // there must be no additional context parameters set on the matched
+    // resource.
+    ContextParams exact_context = 5;
+
+    // .. space reserved for future potential matchers, e.g. CEL expressions.
+  }
+
+  // Directives provide information to data-plane load balancers on how UDPA
+  // resource names are to be resolved. For example, they may provide
+  // alternative resource locators for when primary resolution fails. Directives
+  // are not part of resource names and do not appear in a xDS transport discovery
+  // request.
+  message Directive {
+    oneof directive {
+      option (validate.required) = true;
+
+      // An alternative resource locator for fallback if the resource is
+      // unavailable. For example, take the resource locator:
+      //
+      //   udpa://foo/some-type/some-route-table#alt=udpa://bar/some-type/another-route-table.
+      //
+      // If the data-plane load balancer is unable to reach `foo` to fetch the
+      // resource, it will fallback to `bar`. Alternative resources do not need
+      // to have equivalent content, but they should be functional substitutes.
+      ResourceLocator alt = 1;
+
+      // List collections support inlining of resources via the entry field in
+      // Resource. These inlined Resource objects may have an optional name
+      // field specified. When specified, the entry directive allows
+      // UdpaResourceLocator to directly reference these inlined resources, e.g.
+      // udpa://.../foo#entry=bar.
+      string entry = 2 [(validate.rules).string = {min_len: 1, pattern: "^[0-9a-zA-Z_\\-\\./~:]+$"}];
+    }
+  }
+
+  repeated Directive directives = 6;
```

Comment:
Note that we still have an open thread on how this interacts with RFC3986 URIs with only a single #fragment in the design doc (@markdroth @louiscryan), see https://docs.google.com/document/d/1zZav-IYxMO0mP2A7y5XHBa9v0eXyirw1CxrLLSUqWR8/edit?disco=AAAAGpGgj2A.